keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,987 @@
+
+If Kubernetes wishes to support this format, users workflows will break upon
+upgrade until the offending resources and identified and repaired.
+
+With this CRD Validation Ratcheting feature, Kubernetes would be able to
+adopt new format strings safely without breaking user workflows.
+
+#### K8s Update Widens CRD Validation
+
+In this case the user has a CRD with a field with following schema:
+
+```yaml
+type: string
+format: byte
+```
+
+In k8s 1.27 and below, this format disallows empty strings due to an implementation detail. However, in native types like `Secret`, byte is disallowed. Is Kubernetes had an update to CRDs to allow empty string for `byte`, this widens the validation for those CRDS which use it.
+
+This is fine when upgrading to new versions if Kubernetes: all existing resources still pass validation.
+
+The user may still experience issues when downgrading, for example if they had unreletated issues with the new Kubernetes update:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers fail to update some resources, since they fail validation.
+
+If the CRD Validation Ratcheting feature were enabled the following occurs:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers are allowed to update their resources, as long as all their changes are valid.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+This KEP is scoped to CRDs, but if [KEP-XXX: Declarative Validation]() is implemented then we should expect it to also affect native type validation.
+
+### Risks and Mitigations
+
+Help Wanted
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The basic idea is to add a post-processing step to the `Validate(Update)` methods of the CRD strategy. Given a list of `StatusError`, we check the field path associated with the error, if it is unchanged from old to new then the error is changed into a warning.
```

Comment:
Today, all CEL expressions report validation errors using a field path that works for the plan described here. 

But this might not play nice with the `fieldPath` proposed in https://github.com/kubernetes/kubernetes/pull/118041 since this would allow CEL to report a validation error on a specific field, but CEL might have observed other fields when determining that there was a validation error.

cc @cici37

Reply:
Added a note that we will have to use the field which contains the rule in its `x-kubernetes-validations` and will likely need new plumbing to support that